File: backend/sdk/client/stream_parser.py
"""StreamParser - Parses Claude Agent SDK streaming messages using typed message classes."""

import logging
from dataclasses import dataclass, field
from typing import Optional

from claude_agent_sdk.types import (
    AssistantMessage,
    ResultMessage,
    StreamEvent,
    SystemMessage,
    TextBlock,
    ThinkingBlock,
    ToolUseBlock,
)

logger = logging.getLogger(__name__)


@dataclass
class ParsedStreamMessage:
    """Structured result from parsing SDK stream messages.

    Attributes:
        response_text: Accumulated response text (previous + new delta)
        thinking_text: Accumulated thinking text (previous + new delta)
        session_id: Session ID if found in this message, None otherwise
        skip_used: True if skip tool was called in this message
        memory_entries: List of new memory entries from this message
        anthropic_calls: List of anthropic tool call arguments from this message
        structured_output: Structured output data if using output_format (e.g., WorldSeed)
        usage: Token usage data from ResultMessage (input_tokens, output_tokens, cache info)
        tool_start_name: Tool name from content_block_start (tool_use), None otherwise
        tool_input_delta: Partial JSON from input_json_delta, None otherwise
        content_block_stopped: True if content_block_stop event
    """

    response_text: str
    thinking_text: str
    session_id: Optional[str] = None
    skip_used: bool = False
    memory_entries: list[str] = field(default_factory=list)
    anthropic_calls: list[str] = field(default_factory=list)
    structured_output: Optional[dict] = None
    usage: Optional[dict] = None
    tool_start_name: Optional[str] = None
    tool_input_delta: Optional[str] = None
    content_block_stopped: bool = False

    @property
    def has_tool_usage(self) -> bool:
        """Check if any tools were used in this message."""
        return self.skip_used or bool(self.memory_entries) or bool(self.anthropic_calls)


class StreamParser:
    """Parses Claude Agent SDK streaming messages using typed message classes.

    This class is stateless - all methods can be static or class methods.
    """

    @staticmethod
    def parse_message(message, current_response: str, current_thinking: str) -> ParsedStreamMessage:
        """Parse a streaming message from Claude Agent SDK.

        Args:
            message: SDK message (AssistantMessage, SystemMessage, StreamEvent, etc.)
            current_response: Accumulated response text so far
            current_thinking: Accumulated thinking text so far

        Returns:
            ParsedStreamMessage with extracted fields and updated accumulated text
        """
        if isinstance(message, StreamEvent):
            return StreamParser._parse_stream_event(message, current_response, current_thinking)

        content_delta = ""
        thinking_delta = ""
        new_session_id = None
        skip_tool_called = False
        memory_entries: list[str] = []
        anthropic_calls: list[str] = []
        structured_output = None
        usage = None

        if isinstance(message, ResultMessage):
            if message.usage:
                usage = message.usage
                logger.info(f"ResultMessage.usage raw: {type(usage).__name__} = {usage}")
            if message.structured_output:
                structured_output = message.structured_output
                logger.debug(f"Extracted structured_output: {type(structured_output)}")

        if isinstance(message, SystemMessage):
            if "session_id" in message.data:
                new_session_id = message.data["session_id"]
                logger.debug(f"Extracted session_id: {new_session_id}")

        if isinstance(message, AssistantMessage):
            for block in message.content:
                if isinstance(block, ToolUseBlock):
                    if block.name.endswith("__skip"):
                        skip_tool_called = True
                        logger.info("Agent chose to skip")
                    elif block.name.endswith("__memorize"):
                        memory_entry = block.input.get("memory_entry", "")
                        if memory_entry:
                            memory_entries.append(memory_entry)
                            logger.info(f"Agent recorded memory: {memory_entry}")
                    elif block.name.endswith("__anthropic"):
                        situation = block.input.get("situation", "")
                        if situation:
                            anthropic_calls.append(situation)
                            logger.info(f"Agent called anthropic tool: {situation}")
                elif isinstance(block, ThinkingBlock):
                    # Skip if already accumulated from StreamEvent deltas
                    if not current_thinking:
                        thinking_delta = block.thinking
                elif isinstance(block, TextBlock):
                    # Skip if already accumulated from StreamEvent deltas.
                    # AssistantMessage/ResultMessage TextBlocks contain the complete
                    # turn text, not an incremental delta - appending would duplicate.
                    if not current_response:
                        content_delta += block.text

        return ParsedStreamMessage(
            response_text=current_response + content_delta,
            thinking_text=current_thinking + thinking_delta,
            session_id=new_session_id,
            skip_used=skip_tool_called,
            memory_entries=memory_entries,
            anthropic_calls=anthropic_calls,
            structured_output=structured_output,
            usage=usage,
        )

    @staticmethod
    def _parse_stream_event(message: StreamEvent, current_response: str, current_thinking: str) -> ParsedStreamMessage:
        """Parse a StreamEvent containing raw Claude API streaming events.

        StreamEvent.event is a raw Claude API event dict. Handles:
        - content_block_delta with text_delta or thinking_delta (text streaming)
        - content_block_delta with input_json_delta (tool input streaming)
        - content_block_start with tool_use type (tool invocation start)
        - content_block_stop (content block completion)
        """
        event = message.event
        content_delta = ""
        thinking_delta = ""
        tool_start_name = None
        tool_input_delta = None
        content_block_stopped = False

        event_type = event.get("type")

        if event_type == "content_block_delta":
            delta = event.get("delta", {})
            delta_type = delta.get("type", "")

            if delta_type == "text_delta":
                content_delta = delta.get("text", "")
            elif delta_type == "thinking_delta":
                thinking_delta = delta.get("thinking", "")
            elif delta_type == "input_json_delta":
                tool_input_delta = delta.get("partial_json", "")

        elif event_type == "content_block_start":
            content_block = event.get("content_block", {})
            if content_block.get("type") == "tool_use":
                tool_start_name = content_block.get("name", "")

        elif event_type == "content_block_stop":
            content_block_stopped = True

        return ParsedStreamMessage(
            response_text=current_response + content_delta,
            thinking_text=current_thinking + thinking_delta,
            session_id=message.session_id if message.session_id and not current_response else None,
            tool_start_name=tool_start_name,
            tool_input_delta=tool_input_delta,
            content_block_stopped=content_block_stopped,
        )


class NarrationStreamExtractor:
    """Extracts the 'narrative' field value incrementally from streaming JSON tool input.

    The narration tool receives JSON like: {"narrative": "Once upon a time...", "options": [...]}
    This class accumulates partial_json fragments and extracts the narrative string
    value as it streams in, handling JSON escape sequences.
    """

    def __init__(self):
        self._buffer = ""
        self._in_narrative = False
        self._emitted = 0  # chars already returned from narrative value
        self._narrative_value = ""  # decoded narrative so far

    def feed(self, partial_json: str) -> str:
        """Feed a partial JSON fragment and return any new narrative text.

        Args:
            partial_json: A fragment of the JSON tool input

        Returns:
            New narrative text delta (empty string if nothing new yet)
        """
        self._buffer += partial_json

        if not self._in_narrative:
            # Look for the "narrative" key pattern: "narrative": "
            # Also handle "narrative" : " (with optional whitespace)
            marker = '"narrative"'
            idx = self._buffer.find(marker)
            if idx == -1:
                return ""
            # Find the colon and opening quote after the key
            rest = self._buffer[idx + len(marker):]
            # Skip whitespace and colon
            colon_idx = rest.find(":")
            if colon_idx == -1:
                return ""
            after_colon = rest[colon_idx + 1:].lstrip()
            if not after_colon or after_colon[0] != '"':
                return ""
            # Found opening quote - start extracting
            self._in_narrative = True
            # Re-position buffer to just after the opening quote
            full_offset = idx + len(marker) + colon_idx + 1 + (len(rest[colon_idx + 1:]) - len(after_colon)) + 1
            self._buffer = self._buffer[full_offset:]

        # Now extract the string value, handling JSON escapes
        return self._extract_narrative_delta()

    def _extract_narrative_delta(self) -> str:
        """Extract new narrative text from buffer, handling JSON escape sequences."""
        new_text = ""
        i = 0
        while i < len(self._buffer):
            ch = self._buffer[i]
            if ch == '"':
                # End of the JSON string value
                self._buffer = self._buffer[i + 1:]
                break
            elif ch == '\\':
                if i + 1 >= len(self._buffer):
                    # Incomplete escape - keep in buffer for next feed
                    self._buffer = self._buffer[i:]
                    break
                esc = self._buffer[i + 1]
                if esc == 'n':
                    new_text += '\n'
                    i += 2
                elif esc == 't':
                    new_text += '\t'
                    i += 2
                elif esc == '"':
                    new_text += '"'
                    i += 2
                elif esc == '\\':
                    new_text += '\\'
                    i += 2
                elif esc == '/':
                    new_text += '/'
                    i += 2
                elif esc == 'r':
                    new_text += '\r'
                    i += 2
                elif esc == 'b':
                    new_text += '\b'
                    i += 2
                elif esc == 'f':
                    new_text += '\f'
                    i += 2
                elif esc == 'u':
                    if i + 5 < len(self._buffer):
                        hex_str = self._buffer[i + 2:i + 6]
                        try:
                            new_text += chr(int(hex_str, 16))
                        except ValueError:
                            new_text += '\\u' + hex_str
                        i += 6
                    else:
                        # Incomplete unicode escape - keep in buffer
                        self._buffer = self._buffer[i:]
                        break
                else:
                    new_text += esc
                    i += 2
            else:
                new_text += ch
                i += 1
        else:
            # Consumed entire buffer
            self._buffer = ""

        self._narrative_value += new_text
        return new_text

File: backend/src/sdk/client/stream-parser.ts
/**
 * Parses Claude Agent SDK streaming messages into a flat, accumulated snapshot.
 *
 * Port of `backend/sdk/client/stream_parser.py`.
 *
 * `parseMessage` is stateless: the caller owns the accumulators and passes them
 * back in on every message. The returned `responseText` / `thinkingText` are the
 * FULL accumulated strings (previous + this message's delta), so a caller that
 * wants a delta slices it off the previous value.
 */

import type { SDKMessage } from '@anthropic-ai/claude-agent-sdk';

/**
 * Token accounting lifted off the `result` message.
 *
 * Normalized to the four wire field names the perf logger reads, with 0 for
 * anything the CLI omitted (Python read these off the raw usage dict with
 * `.get(..., 0)` at the call site).
 */
export interface ParsedUsage {
  input_tokens: number;
  output_tokens: number;
  cache_creation_input_tokens: number;
  cache_read_input_tokens: number;
}

/** Structured result of parsing a single SDK stream message. */
export interface ParsedStreamMessage {
  /** Accumulated response text (previous + this message's delta). */
  responseText: string;
  /** Accumulated thinking text (previous + this message's delta). */
  thinkingText: string;
  /** Session id if this message carried one, else undefined. */
  sessionId: string | undefined;
  /** True if a `*__skip` tool was called in this message. */
  skipUsed: boolean;
  /** New memory entries from `*__memorize` calls in this message. */
  memoryEntries: string[];
  /** New situations from `*__anthropic` calls in this message. */
  anthropicCalls: string[];
  /** Structured output payload from a `result` message, if any. */
  structuredOutput: unknown;
  /** Token usage from a `result` message, if any. */
  usage: ParsedUsage | undefined;
  /** Tool name from a `content_block_start` of type `tool_use`. */
  toolStartName: string | undefined;
  /** Partial JSON from an `input_json_delta`; `''` is meaningful, undefined means "no delta". */
  toolInputDelta: string | undefined;
  /** True if this message was a `content_block_stop` event. */
  contentBlockStopped: boolean;
  /** Convenience: skip, memorize or anthropic was used in this message. */
  hasToolUsage: boolean;
}

/**
 * What `parseMessage` accepts.
 *
 * The union with `Record<string, unknown>` is deliberate: this code runs against
 * whatever CLI/SDK version happens to be installed, so every field is narrowed
 * structurally rather than by matching the `SDKMessage` union exhaustively. An
 * unseen message subtype falls through to "nothing extracted" instead of
 * crashing, and hand-built fixtures need no casts.
 */
export type ParsableMessage = SDKMessage | Record<string, unknown>;

function asRecord(value: unknown): Record<string, unknown> | undefined {
  return typeof value === 'object' && value !== null && !Array.isArray(value)
    ? (value as Record<string, unknown>)
    : undefined;
}

function asString(value: unknown): string | undefined {
  return typeof value === 'string' ? value : undefined;
}

function asArray(value: unknown): readonly unknown[] | undefined {
  return Array.isArray(value) ? (value as readonly unknown[]) : undefined;
}

function asCount(value: unknown): number {
  return typeof value === 'number' && Number.isFinite(value) ? value : 0;
}

interface ParsedFields {
  responseText: string;
  thinkingText: string;
  sessionId?: string | undefined;
  skipUsed?: boolean;
  memoryEntries?: string[];
  anthropicCalls?: string[];
  structuredOutput?: unknown;
  usage?: ParsedUsage | undefined;
  toolStartName?: string | undefined;
  toolInputDelta?: string | undefined;
  contentBlockStopped?: boolean;
}

function finalize(fields: ParsedFields): ParsedStreamMessage {
  const skipUsed = fields.skipUsed ?? false;
  const memoryEntries = fields.memoryEntries ?? [];
  const anthropicCalls = fields.anthropicCalls ?? [];
  return {
    responseText: fields.responseText,
    thinkingText: fields.thinkingText,
    sessionId: fields.sessionId,
    skipUsed,
    memoryEntries,
    anthropicCalls,
    structuredOutput: fields.structuredOutput,
    usage: fields.usage,
    toolStartName: fields.toolStartName,
    toolInputDelta: fields.toolInputDelta,
    contentBlockStopped: fields.contentBlockStopped ?? false,
    hasToolUsage: skipUsed || memoryEntries.length > 0 || anthropicCalls.length > 0,
  };
}

/**
 * Parse one `stream_event` message (an `SDKPartialAssistantMessage`, emitted only
 * when the query was started with `includePartialMessages: true`).
 *
 * `event` is a raw Anthropic Messages API streaming event.
 */
function parseStreamEvent(
  message: Record<string, unknown>,
  currentResponse: string,
  currentThinking: string,
): ParsedStreamMessage {
  const event = asRecord(message.event) ?? {};
  const eventType = asString(event.type);

  let contentDelta = '';
  let thinkingDelta = '';
  let toolStartName: string | undefined;
  let toolInputDelta: string | undefined;
  let contentBlockStopped = false;

  if (eventType === 'content_block_delta') {
    const delta = asRecord(event.delta) ?? {};
    const deltaType = asString(delta.type) ?? '';

    if (deltaType === 'text_delta') {
      contentDelta = asString(delta.text) ?? '';
    } else if (deltaType === 'thinking_delta') {
      thinkingDelta = asString(delta.thinking) ?? '';
    } else if (deltaType === 'input_json_delta') {
      // '' rather than undefined: callers distinguish "no tool input this
      // message" (undefined) from "an empty fragment" (''), and the SDK does
      // emit empty `partial_json` fragments.
      toolInputDelta = asString(delta.partial_json) ?? '';
    }
  } else if (eventType === 'content_block_start') {
    const contentBlock = asRecord(event.content_block) ?? {};
    if (asString(contentBlock.type) === 'tool_use') {
      toolStartName = asString(contentBlock.name) ?? '';
    }
  } else if (eventType === 'content_block_stop') {
    contentBlockStopped = true;
  }

  const sessionId = asString(message.session_id);

  return finalize({
    responseText: currentResponse + contentDelta,
    thinkingText: currentThinking + thinkingDelta,
    // Only report the session id before any response text exists. Every
    // stream_event carries it, so reporting it unconditionally would make the
    // caller re-assign the session on every token.
    sessionId: sessionId && !currentResponse ? sessionId : undefined,
    toolStartName,
    toolInputDelta,
    contentBlockStopped,
  });
}

/**
 * Parse a streaming message from the Claude Agent SDK.
 *
 * @param message - Any SDK message; unrecognized shapes yield an inert result.
 * @param currentResponse - Response text accumulated so far.
 * @param currentThinking - Thinking text accumulated so far.
 */
export function parseMessage(
  message: ParsableMessage,
  currentResponse: string,
  currentThinking: string,
): ParsedStreamMessage {
  const record = asRecord(message);
  if (!record) {
    return finalize({ responseText: currentResponse, thinkingText: currentThinking });
  }

  const messageType = asString(record.type);

  if (messageType === 'stream_event') {
    return parseStreamEvent(record, currentResponse, currentThinking);
  }

  let contentDelta = '';
  let thinkingDelta = '';
  let sessionId: string | undefined;
  let skipUsed = false;
  const memoryEntries: string[] = [];
  const anthropicCalls: string[] = [];
  let structuredOutput: unknown;
  let usage: ParsedUsage | undefined;

  if (messageType === 'result') {
    const rawUsage = asRecord(record.usage);
    // An empty usage object counts as "no usage", matching the Python truthiness check.
    if (rawUsage && Object.keys(rawUsage).length > 0) {
      usage = {
        input_tokens: asCount(rawUsage.input_tokens),
        output_tokens: asCount(rawUsage.output_tokens),
        cache_creation_input_tokens: asCount(rawUsage.cache_creation_input_tokens),
        cache_read_input_tokens: asCount(rawUsage.cache_read_input_tokens),
      };
    }
    if (record.structured_output !== undefined && record.structured_output !== null) {
      structuredOutput = record.structured_output;
    }
  }

  if (messageType === 'system') {
    // The TS SDK puts session_id at the top level; the Python SDK nested it
    // under `data`. Check both so either CLI shape is handled.
    sessionId = asString(record.session_id) ?? asString(asRecord(record.data)?.session_id);
  }

  if (messageType === 'assistant') {
    // TS SDK: blocks live on `message.message.content` (an Anthropic Message).
    // Python SDK: they live on `message.content`.
    const content =
      asArray(asRecord(record.message)?.content) ?? asArray(record.content) ?? [];

    for (const rawBlock of content) {
      const block = asRecord(rawBlock);
      if (!block) continue;
      const blockType = asString(block.type);

      if (blockType === 'tool_use') {
        // Tools are namespaced per agent (`<agent>__skip`), so match by suffix.
        const name = asString(block.name) ?? '';
        const input = asRecord(block.input) ?? {};

        if (name.endsWith('__skip')) {
          skipUsed = true;
        } else if (name.endsWith('__memorize')) {
          const memoryEntry = asString(input.memory_entry) ?? '';
          if (memoryEntry) memoryEntries.push(memoryEntry);
        } else if (name.endsWith('__anthropic')) {
          const situation = asString(input.situation) ?? '';
          if (situation) anthropicCalls.push(situation);
        }
      } else if (blockType === 'thinking') {
        // Load-bearing: an assistant message's thinking block holds the WHOLE
        // turn's thinking, not an increment. If stream_event deltas already
        // accumulated it, appending here would emit it twice.
        if (!currentThinking) thinkingDelta += asString(block.thinking) ?? '';
      } else if (blockType === 'text') {
        // Same rule as thinking: assistant/result text blocks carry the complete
        // turn text, so they are only used when nothing was streamed.
        if (!currentResponse) contentDelta += asString(block.text) ?? '';
      }
    }
  }

  return finalize({
    responseText: currentResponse + contentDelta,
    thinkingText: currentThinking + thinkingDelta,
    sessionId,
    skipUsed,
    memoryEntries,
    anthropicCalls,
    structuredOutput,
    usage,
  });
}
